src/common/ceph_crypto.cc

```diff
@@ -16,77 +16,140 @@
+static size_t ssl_num_locks;
+#endif /* not OPENSSL_API_1_1 */
+
+#ifndef OPENSSL_API_1_1
+static void
+ssl_locking_callback(int mode, int mutex_num, const char *file, int line)
 {
-  pid_t pid = getpid();
-  pthread_mutex_lock(&crypto_init_mutex);
-  if (crypto_init_pid != pid) {
-    if (crypto_init_pid > 0) {
-      SECMOD_RestartModules(PR_FALSE);
-    }
-    crypto_init_pid = pid;
+  static_cast<void>(line);
+  static_cast<void>(file);
+
+  if (mode & 1) {
+    /* 1 is CRYPTO_LOCK */
+    [[maybe_unused]] auto r = pthread_mutex_lock(&ssl_mutexes[mutex_num]);
+  } else {
+    [[maybe_unused]] auto r = pthread_mutex_unlock(&ssl_mutexes[mutex_num]);
   }
+}
 
-  if (++crypto_refs == 1) {
-    NSSInitParameters init_params;
-    memset(&init_params, 0, sizeof(init_params));
-    init_params.length = sizeof(init_params);
+static unsigned long
+ssl_get_thread_id(void)
+{
+  static_assert(sizeof(unsigned long) >= sizeof(pthread_t));
+  /* pthread_t may be any data type, so a simple cast to unsigned long
+   * can rise a warning/error, depending on the platform.
+   * Here memcpy is used as an anything-to-anything cast. */
+  unsigned long ret = 0;
+  pthread_t t = pthread_self();
+  memcpy(&ret, &t, sizeof(pthread_t));
+  return ret;
+}
+#endif /* not OPENSSL_API_1_1 */
 
-    uint32_t flags = (NSS_INIT_READONLY | NSS_INIT_PK11RELOAD);
-    if (cct->_conf->nss_db_path.empty()) {
-      flags |= (NSS_INIT_NOCERTDB | NSS_INIT_NOMODDB);
-    }
-    crypto_context = NSS_InitContext(cct->_conf->nss_db_path.c_str(), "", "",
-                                     SECMOD_DB, &init_params, flags);
+static void init() {
+#ifndef OPENSSL_API_1_1
+  if (++crypto_refs != 1) {
+    return;
   }
-  pthread_mutex_unlock(&crypto_init_mutex);
-  ceph_assert_always(crypto_context != NULL);
-}
 
-void ceph::crypto::shutdown(bool shared)
-{
-  pthread_mutex_lock(&crypto_init_mutex);
-  ceph_assert_always(crypto_refs > 0);
-  if (--crypto_refs == 0) {
-    NSS_ShutdownContext(crypto_context);
-    if (!shared) {
-      PR_Cleanup();
+  /* According to https://wiki.openssl.org/index.php/Library_Initialization#libcrypto_Initialization */
+  OpenSSL_add_all_algorithms();
+  ERR_load_crypto_strings();
+
+  /* Initialize locking callbacks, needed for thread safety.
+   * http://www.openssl.org/support/faq.html#PROG1
+   */
+  ssl_num_locks = std::max(CRYPTO_num_locks(), 0);
+  if (ssl_num_locks > 0) {
+    try {
+      ssl_mutexes = new pthread_mutex_t[ssl_num_locks];
+    } catch (...) {
+      ceph_assert_always("can't allocate memory for OpenSSL init" == nullptr);
     }
-    crypto_context = NULL;
-    crypto_init_pid = 0;
   }
-  pthread_mutex_unlock(&crypto_init_mutex);
+
+  pthread_mutexattr_t pthread_mutex_attr;
+  pthread_mutexattr_init(&pthread_mutex_attr);
+  pthread_mutexattr_settype(&pthread_mutex_attr, PTHREAD_MUTEX_RECURSIVE);
+
+  for (size_t i = 0; i < ssl_num_locks; i++) {
+    pthread_mutex_init(&ssl_mutexes[i], &pthread_mutex_attr);
```

Comment:
we could use `ceph::make_shared_mutex ` to further simplify this (or continue using `ceph::recursive_mutex` if we think plain mutex is good enough:
```c++
static vector<ceph::make_shared_mutex > ssl_mutexes;
// ...
static void init() {
  // ...
  std::transform(boost::make_counting_iterator(0),
                 boost::make_counting_iterator(ssl_num_locks),
                 back_inserter(ssl_mutexes), [](int i) {
                   return ceph::make_shared_mutex{fmt::format("openssl.{}", i)};
                 });
  // ...
}

static void ssl_locking_callback(int mode, int n, const char *, int) {
  static constexpr int CRYPTO_LOCK = 0b0001;
  static constexpr int CRYPTO_UNLOCK = 0b0010;
  static constexpr int CRYPTO_READ = 0b0100;
  static constexpr int CRYPTO_WRITE = 0b1000;
  if (n < 0 || n >= ssl_num_locks) {
    return;
  }
  if (mode & CRYPTO_READ) {
    if (mode & CRYPTO_LOCK) {
      ssl_mutexes[mutex_num].lock_shared();
    } else if (mode & CRYPTO_UNLOCK) {
      ssl_mutexes[mutex_num].unlock_shared();
    }
  } else if (mode & CRYPTO_WRITE) {
    if (mode & CRYPTO_LOCK) {
      ssl_mutexes[mutex_num].lock();
    } else if (mode & CRYPTO_UNLOCK) {
      ssl_mutexes[mutex_num].unlock();
    }
  }
}
```

Reply:
Moved to the vector. Thanks for the snippet, @tchaikov!